### PR TITLE
Publicize required methods to allow for custom implementations of resource detectors.

### PR DESCRIPTION
**Pull request by Yun-Ting**
## Changes
Addressed https://github.com/open-telemetry/opentelemetry-dotnet/pull/2897#issuecomment-1040385720
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2897 alone itself was not sufficient for the users to implement their own resource detector. And this PR bridge this gap.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [x] Changes in public API reviewed

### CODE REVIEW

File: docs/trace/extending-the-sdk/MyResourceDetector.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using OpenTelemetry.Resources;
+
+internal class MyResourceDetector : IResourceDetector
```
Comment (reyang) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:21]: How is this being used in this example?

File: docs/trace/extending-the-sdk/MyResourceDetector.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public Resource Detect()
+    {
```
Comment (reyang) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:24]: Might need to update the docs https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk#extending-the-opentelemetry-net-sdk

File: docs/trace/extending-the-sdk/MyResourceDetector.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        Console.WriteLine("Implement customized detection behavior for MyResourceDector:");
```
Comment (reyang) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:25]: The message seems a bit weird.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        var attributes = new List<KeyValuePair<string, object>>
+        {
+            new KeyValuePair<string, object>("key", "val"),
+        };
+
+        Resource resource = new(attributes);
```
Comment (reyang) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:31]: nit: might be simplified to `return new Resource(attributes)`.

File: src/OpenTelemetry/Resources/Resource.cs [hunk 5]
```diff
@@ -35,7 +35,7 @@ public class Resource
         /// Initializes a new instance of the <see cref="Resource"/> class.
         /// </summary>
         /// <param name="attributes">An <see cref="IEnumerable{T}"/> of attributes that describe the resource.</param>
-        internal Resource(IEnumerable<KeyValuePair<string, object>> attributes)
+        public Resource(IEnumerable<KeyValuePair<string, object>> attributes)
```
Comment (cijothomas) [inline src/OpenTelemetry/Resources/Resource.cs:38]: please keep this internal as before.
We already exposed IResourceDetector, and we are missing a public method to add detectors : https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Resources/ResourceBuilder.cs#L89-L91 --> this is the thing to public ize to address the issue.

**Review by cijothomas — CHANGES_REQUESTED** [review]
See https://github.com/open-telemetry/opentelemetry-dotnet/pull/2949/files#r815222141

File: docs/trace/extending-the-sdk/MyResourceDetector.cs (hunk 1 above)
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs:21]: Added usage.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs (hunk 3 above)
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs]: I've removed it.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs (hunk 4 above)
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs]: I've done a minor refactor to the detect method.

File: src/OpenTelemetry/Resources/Resource.cs (hunk 5 above)
Comment (cijothomas) [reply src/OpenTelemetry/Resources/Resource.cs:38]: After re-reading, and offline discussion, we'll have to expose this public, as Detectors need to return Resource, so need public ctor to do that.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs (hunk 2 above)
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs:26]: Update the file to point to the example.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+using OpenTelemetry.Resources;
+
+internal class MyResourceDetector : IResourceDetector
+{
+    public const string EnVarkey = "myEnVarkey";
+
+    public Resource Detect()
+    {
+        var resource = Resource.Empty;
+        if (this.LoadString(EnVarkey, out string envAttributeVal))
```
Comment (cijothomas) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:28]: keep this as simple as possible. I'd recommend to simply return a new List of attributes. Don't need to show env variable parsing or anything here.

File: docs/trace/extending-the-sdk/MyResourceDectectorExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using OpenTelemetry.Resources;
+
+internal static class MyResourceDectectorExtensions
```
Comment (cijothomas) [inline docs/trace/extending-the-sdk/MyResourceDectectorExtensions.cs:19]: not really required, right? I'd skip this whole class.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs (hunk 6 above)
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs:28]: I think it serves as a good example for the users to provide a more wholistic usage on where to get the resource attributes, at lease for me I think it is good to have information. But I'm open to other suggestions.

File: docs/trace/extending-the-sdk/MyResourceDectectorExtensions.cs (hunk 7 above)
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDectectorExtensions.cs]: removed

File: docs/trace/extending-the-sdk/MyResourceDetector.cs (hunk 6 above)
Comment (cijothomas) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs:28]: eg: Sampler (doc has the simplest one, but links to other official ones from this repo)

**Review by cijothomas — APPROVED** [review]
LGTM.
Will wait to get one more approval before merge, as this is adding new public API very close to a stable release, and want to get one more review.

**Review by reyang — APPROVED** [review]
LGTM.

File: docs/trace/extending-the-sdk/README.md [hunk 8]
```diff
@@ -4,6 +4,7 @@
 * [Building your own instrumentation library](#instrumentation-library)
 * [Building your own processor](#processor)
 * [Building your own sampler](#sampler)
+* [Building your own resourceDetector](#resourceDetector)
```
Comment (reyang) [inline docs/trace/extending-the-sdk/README.md:7]:
```suggestion
* [Building your own resource detector](#resourcedetector)
```

File: docs/trace/extending-the-sdk/README.md [hunk 9]
```diff
@@ -293,6 +294,17 @@ class MySampler : Sampler
 
 A demo sampler is shown [here](./MySampler.cs).
 
+## ResourceDetector
+
+Custom ResourceDetector can be implemented:
```
Comment (alanwest) [inline docs/trace/extending-the-sdk/README.md:299]:
```suggestion
## Resource Detector

OpenTelemetry .NET SDK provides a resource detector for detecting
resource information from the `OTEL_RESOURCE_ATTRIBUTES` and
`OTEL_SERVICE_NAME` environment variables.

Custom resource detectors can be implemented:
```

**Review by alanwest — APPROVED** [review]
LGTM

**Review by mic-max — APPROVED** [review]

**Review by cijothomas — APPROVED** [review]

**Pull request merged**